File: src/models/Order.ts
import mongoose, { Schema, models, model, Types } from "mongoose";

export type DeliveryStatus =
  | "processing"
  | "packed"
  | "shipped"
  | "out_for_delivery"
  | "delivered"
  | "cancelled"
  | "returned";

export type PaymentStatus = "pending" | "paid" | "failed";

export type TrackingProvider =
  | "delhivery"
  | "bluedart"
  | "dtdc"
  | "indiapost"
  | "shiprocket"
  | "other";

export interface IOrderItem {
  productId: string;
  name: string;
  price: number;
  quantity: number;
  size: string;
  image: string;
  category?: string;
}

export interface IShippingAddress {
  name: string;
  email: string;
  phone: string;
  address: string;
  city: string;
  state: string;
  pincode: string;
}

export interface IOrder {
  userId: Types.ObjectId;
  orderNumber: string;
  items: IOrderItem[];
  shippingAddress: IShippingAddress;
  subtotal: number;
  shipping: number;
  total: number;
  currency: string;
  paymentStatus: PaymentStatus;
  deliveryStatus: DeliveryStatus;
  trackingProvider?: TrackingProvider;
  trackingNumber?: string;
  trackingUrl?: string;
  shippedAt?: Date;
  deliveredAt?: Date;
  razorpayOrderId?: string;
  razorpayPaymentId?: string;
}

const OrderSchema = new Schema<IOrder>(
  {
    userId: { type: Schema.Types.ObjectId, ref: "User", required: true },
    orderNumber: { type: String, required: true, unique: true },
    items: [
      {
        productId: String,
        name: String,
        price: Number,
        quantity: Number,
        size: String,
        image: String,
        category: String,
      },
    ],
    shippingAddress: {
      name: String,
      email: String,
      phone: String,
      address: String,
      city: String,
      state: String,
      pincode: String,
    },
    subtotal: { type: Number, required: true },
    shipping: { type: Number, default: 0 },
    total: { type: Number, required: true },
    currency: { type: String, default: "INR" },
    paymentStatus: {
      type: String,
      enum: ["pending", "paid", "failed"],
      default: "pending",
    },
    deliveryStatus: {
      type: String,
      enum: [
        "processing",
        "packed",
        "shipped",
        "out_for_delivery",
        "delivered",
        "cancelled",
        "returned",
      ],
      default: "processing",
    },
    trackingProvider: {
      type: String,
      enum: ["delhivery", "bluedart", "dtdc", "indiapost", "shiprocket", "other"],
    },
    trackingNumber: String,
    trackingUrl: String,
    shippedAt: Date,
    deliveredAt: Date,
    razorpayOrderId: String,
    razorpayPaymentId: String,
  },
  { timestamps: true }
);

OrderSchema.index({ deliveryStatus: 1, createdAt: -1 });
OrderSchema.index({ userId: 1 });

export default models.Order || model<IOrder>("Order", OrderSchema);

File: mizhara-backend/models/order.go
package models

import "time"

type DeliveryStatus string
type PaymentStatus string
type TrackingProvider string

const (
	DeliveryProcessing DeliveryStatus = "processing"
	DeliveryShipped    DeliveryStatus = "shipped"
	DeliveryDelivered  DeliveryStatus = "delivered"
	PaymentPending     PaymentStatus  = "pending"
	PaymentPaid        PaymentStatus  = "paid"
	PaymentFailed      PaymentStatus  = "failed"
)

type OrderItem struct {
	ProductID string  `json:"productId"`
	Name      string  `json:"name"`
	Price     float64 `json:"price"`
	Quantity  int     `json:"quantity"`
	Size      string  `json:"size"`
	Image     string  `json:"image"`
	Category  string  `json:"category,omitempty"`
}

type ShippingAddress struct {
	Name    string `json:"name"`
	Email   string `json:"email"`
	Phone   string `json:"phone"`
	Address string `json:"address"`
	City    string `json:"city"`
	State   string `json:"state"`
	Pincode string `json:"pincode"`
}

type Order struct {
	ID                string           `json:"id"`
	UserID            string           `json:"userId"`
	OrderNumber       string           `json:"orderNumber"`
	Items             []OrderItem      `json:"items"`
	ShippingAddress   ShippingAddress  `json:"shippingAddress"`
	Subtotal          float64          `json:"subtotal"`
	DiscountAmount    float64          `json:"discountAmount,omitempty"`
	OfferID           string           `json:"offerId,omitempty"`
	OfferName         string           `json:"offerName,omitempty"`
	Shipping          float64          `json:"shipping"`
	Total             float64          `json:"total"`
	Currency          string           `json:"currency"`
	PaymentStatus     PaymentStatus    `json:"paymentStatus"`
	DeliveryStatus    DeliveryStatus   `json:"deliveryStatus"`
	TrackingProvider  TrackingProvider `json:"trackingProvider,omitempty"`
	TrackingNumber    string           `json:"trackingNumber,omitempty"`
	TrackingURL       string           `json:"trackingUrl,omitempty"`
	ShippedAt         *time.Time       `json:"shippedAt,omitempty"`
	DeliveredAt       *time.Time       `json:"deliveredAt,omitempty"`
	RazorpayOrderID   string           `json:"razorpayOrderId,omitempty"`
	RazorpayPaymentID string           `json:"razorpayPaymentId,omitempty"`
	CreatedAt         time.Time        `json:"createdAt"`
	UpdatedAt         time.Time        `json:"updatedAt"`
}
